feat(config)!: migrate zebrad to use a layered configuration

gustavovalverde opened the pull request:
### 

## Motivation

- Replace Abscissa-based config loading and the Docker TOML generation flow with a standard layered configuration approach.
- Let users configure via environment variables (`ZEBRA_SECTION__KEY`) without an entrypoint-generated `zebrad.toml`.
- Remove legacy `ZEBRA_*` test/internal env vars to avoid collisions with the new `config-rs` prefix.

## Solution

- Config loading
  - Implement `ZebradConfig::load()` using config-rs (defaults → optional TOML → env).
  - Derive/ensure defaults and `deny_unknown_fields` in config structs.

- Docker entrypoint
  - Remove config generation
  - Keep test execution path for `cargo nextest`.

- Environment variables (breaking)
  - Stop honoring legacy `ZEBRA_*` test/path vars. Use:
    - `SKIP_NETWORK_TESTS`, `SKIP_IPV6_TESTS`
    - `TEST_LIGHTWALLETD`, `TEST_LARGE_CHECKPOINTS`, `TEST_SYNC_TO_CHECKPOINT`, `TEST_SYNC_PAST_CHECKPOINT`
    - `STATE_CACHE_DIR` instead of `ZEBRA_CACHE_DIR`
    - `FORCE_USE_COLOR` instead of `ZEBRA_FORCE_USE_COLOR`
  - Logging: stop using `RUST_LOG`/`ZEBRA_RUST_LOG`; use `ZEBRA_TRACING__FILTER`.

- Tests and helpers
  - Provide a default network-matched `mining.miner_address` in the test config helper, selected by `ZEBRA_NETWORK__NETWORK` or the test’s requested network; still overridable by `ZEBRA_MINING__MINER_ADDRESS` or TOML.
  - Add `zebrad/tests/common/configs/v2.4.2.toml` and normalize `cache_dir`, `[rpc].cookie_dir = "cache_dir"`.

- Documentation
  - Update `generate.rs` docs to show `ZEBRA_SECTION__KEY` usage.

- Breaking changes
  - Legacy `ZEBRA_*` test/path environment variables are removed.
  - Docker entrypoint no longer generates/rewrites TOML; configuration is defaults + optional TOML + `ZEBRA_*` env.

### Tests

Our whole test suite must pass

### Follow-up Work

- Remove any remaining Abscissa `Configurable` trait usage.
- Update Zebra Book docs for the new configuration model.
- Update CI to use new env variable names.
- Validate z3 docker-compose with the new env-only configuration.

### PR Checklist

- [ ] The PR name is suitable for the release notes.
- [ ] The solution is tested.
- [ ] The documentation is up to date.

gustavovalverde committed:
ref(docker): improve cargo caching by aligning mounts with CARGO_HOME

Previously, cache mounts targeted specific subdirectories which did not match
the CARGO_HOME path, leading to ineffective caching and repeated downloads.
This change uses a single cache mount for CARGO_HOME in both test and release
stages, ensuring Cargo can leverage BuildKit's cache for faster rebuilds.

Additionally, introduced CARGO_TARGET_DIR for consistent build output paths,
replaced bind mounts with COPY --link for test stage to ensure source code
availability, and updated cache handling to preserve artifacts across builds.

Fixes #9331
Relates #6169

gustavovalverde committed:
chore: fix typo

gustavovalverde committed:
test: do not use cache mounts

gustavovalverde committed:
chore: trigger a build

gustavovalverde committed:
imp(docker): revert to binds with `-Z checksum-freshness` Rust flag

gustavovalverde committed:
chore: trigger a build

gustavovalverde committed:
chore: reduce diff

gustavovalverde committed:
fix: remove extra steps

gustavovalverde committed:
fix: reduce complexity

gustavovalverde committed:
revert: use a better approach

gustavovalverde committed:
fix: create and own the cargo and cargo target folders

gustavovalverde committed:
test building several features

gustavovalverde committed:
use nextest and avoid rustup compilation at runtime

gustavovalverde committed:
refactor: consolidate test features and migrate to cargo-nextest

- Override FEATURES in Dockerfile tests stage with complete feature set needed for all tests
- Convert entrypoint.sh from cargo test to cargo nextest run with --workspace scope
- Update nextest.toml test filters to match actual function names:
  - generate_checkpoints -> generate_checkpoints_mainnet/testnet
  - fake_activation_heights -> with_fake_activation_heights
- Add nextest overrides for lightwalletd and sync test categories
- Remove feature duplication by using centralized FEATURES env var
- Align build and test scopes between Dockerfile and entrypoint to prevent recompilation

This ensures consistent feature sets across build and test phases while
standardizing on nextest for better test execution and filtering.

gustavovalverde committed:
fix: use correct package management

gustavovalverde committed:
fix: better handle different test cases

gustavovalverde committed:
refactor(ci): use nextest for triggering tests

gustavovalverde committed:
fix: use nextest overrides

gustavovalverde committed:
fix: delete extra profiles

gustavovalverde committed:
fix: use default filter instead

gustavovalverde committed:
refactor(testing): Use envars and nextest profiles for test control

This commit refactors the testing architecture to remove test-specific feature flags in favor of runtime environment variables and nextest profiles. This simplifies the build process, as a single test binary can be built and used for all test scenarios.

The primary changes are:
- Test-specific features like `lightwalletd-grpc-tests`, `sync_*`, and `zebra-checkpoints` have been removed from `Cargo.toml`.
- The tests formerly guarded by these features are now controlled by `ZEBRA_TEST_*` environment variables. These act as opt-in safety guards, preventing developers from accidentally running long or resource-intensive tests locally.
- CI workflows in `.github/workflows` have been updated. They now use `nextest` profiles to select specific tests and set the corresponding `ZEBRA_TEST_*` variables to activate them.
- The `tonic-build` dependency, required for `lightwalletd` tests, is now correctly configured as a test-only build dependency. It is enabled through the `zebra-test` feature, ensuring it is not included in production release builds. This resolves a build warning related to an unknown `cfg` value.

gustavovalverde committed:
refactor: remove compile-time feature gates from tests in favor of runtime controls

- Remove lightwalletd-grpc-tests feature from zebra-test crate
- Move lightwalletd-grpc-tests feature definition in zebrad/Cargo.toml to proper location
- Make tonic-build conditional on lightwalletd-grpc-tests feature in build.rs
- Remove feature gates from test modules, allowing them to compile unconditionally
- Remove feature gate from lightwalletd_test_suite test execution
- Update CI workflow to include lightwalletd-grpc-tests feature for comprehensive testing

Tests now use runtime environment variables (ZEBRA_TEST_*) instead of compile-time
features, allowing a single test binary to run different test suites based on
configuration while keeping production builds lean.

gustavovalverde committed:
fix: build

gustavovalverde committed:
feat: add indexer as a default feature

gustavovalverde committed:
refactor: make lightwalletd-grpc-tests and indexer features optional

- Add feature gates to lightwalletd test infrastructure to prevent compilation errors when lightwalletd-grpc-tests is disabled
- Add feature gate to indexer test to prevent compilation errors when indexer is disabled
- Move lightwalletd-related imports and constants behind feature gates
- Wrap gRPC code generation in feature-conditional module
- Fix GitHub Actions workflow to pass features as single string
- Restore missing lightwalletd_failure_messages method with feature gate
- Add missing DATABASE_FORMAT_UPGRADE_IS_LONG import

This ensures --no-default-features builds work correctly while maintaining full functionality when features are enabled.

gustavovalverde committed:
refactor(tests): remove unused import and use fully qualified path

- Remove unused DATABASE_FORMAT_UPGRADE_IS_LONG import
- Update references to use common::cached_state::DATABASE_FORMAT_UPGRADE_IS_LONG
- Fix unused import linting warning

gustavovalverde committed:
refactor(docker): simplify dockerfile

gustavovalverde committed:
fix: formatting

gustavovalverde committed:
fix: use the default cargo directories

gustavovalverde committed:
fix: build

gustavovalverde committed:
fix: cargo ownership

gustavovalverde committed:
fix: use correct test filtering with nextest

gustavovalverde committed:
fix: avoid cargo cache invalidation

gustavovalverde committed:
fix(ci): Correct Docker test execution and platform filtering

This commit addresses several issues related to running tests within Docker and CI environments.

The initial problem was a permissions error where `nextest` could not write to its store directory. This was resolved by adjusting `CARGO_HOME` and `CARGO_TARGET_DIR` to be relative to the user's home directory within the Docker image.

A subsequent issue was discovered where test filters in `nextest.toml` were platform-specific, causing the entire test suite to run on `x86_64` CI runners, leading to failures. The configuration has been refactored to be platform-agnostic, ensuring filters are applied correctly on all architectures.

Additionally, the `Dockerfile` has been updated to use a multi-stage build for fetching the `lightwalletd` binary, resolving multi-platform build failures. The test entrypoint script was also improved to correctly handle ignored tests and provide cleaner logs.

Finally, the GCP integration test workflow has been simplified to rely on the container's exit code for determining test success, removing fragile log parsing.

gustavovalverde committed:
fix: improve caching

gustavovalverde committed:
fix: segregate sync-past-mandatory test

gustavovalverde committed:
fix: show output for specific test

gustavovalverde committed:
revert: most sync and lwd changes

gustavovalverde committed:
fix: remove indexer

gustavovalverde committed:
chore: reduce diff

gustavovalverde committed:
fix: nextest timeouts

gustavovalverde committed:
Apply suggestions from code review

Co-authored-by: Conrado Gouvea <conrado@zfnd.org>

gustavovalverde committed:
fix: linting errors

gustavovalverde committed:
chore: update documentation and add missing profiles

gustavovalverde committed:
chore: more documentation fixes

gustavovalverde committed:
feat(config)!: migrate zebrad to config-rs; drop legacy env vars; update tests and Docker

- Replace Abscissa-based config load with config-rs layered loading
  - Implement `ZebradConfig::load()` using defaults + TOML + env (prefix `ZEBRA`, `__` for nesting)
  - Integrate in `application.rs` and `commands/copy_state.rs`
- Simplify Docker entrypoint
  - Remove TOML generation and legacy env translation
  - Keep test execution path for `cargo nextest`
- Rename and standardize environment variables (breaking)
  - Tests: `ZEBRA_*` → `TEST_*` or unprefixed:
    - `ZEBRA_SKIP_NETWORK_TESTS` → `SKIP_NETWORK_TESTS`
    - `ZEBRA_SKIP_IPV6_TESTS` → `SKIP_IPV6_TESTS`
    - `ZEBRA_TEST_LIGHTWALLETD` → `TEST_LIGHTWALLETD`
    - `ZEBRA_TEST_LARGE_CHECKPOINTS` → `TEST_LARGE_CHECKPOINTS`
    - `ZEBRA_TEST_SYNC_TO_CHECKPOINT` → `TEST_SYNC_TO_CHECKPOINT`
    - `ZEBRA_TEST_SYNC_PAST_CHECKPOINT` → `TEST_SYNC_PAST_CHECKPOINT`
  - Paths: `ZEBRA_CACHE_DIR` → `STATE_CACHE_DIR`
  - Colors: `ZEBRA_FORCE_USE_COLOR` → `FORCE_USE_COLOR`
  - Logging: stop using `ZEBRA_RUST_LOG`; use `ZEBRA_TRACING__FILTER` (maps to `tracing.filter`)
- Fix acceptance tests and helpers for new config
  - Remove backwards-compat fallbacks for old env names
  - Deduplicate conflicting constants in `cached_state.rs` and `lightwalletd.rs`
  - Provide default network-matched mining address in test config:
    - Selects Mainnet/Testnet/Regtest UA via `ZEBRA_NETWORK__NETWORK` or function param
    - Still overridable via `ZEBRA_MINING__MINER_ADDRESS` or TOML
- Add missing generated config for compatibility tests
  - Commit `zebrad/tests/common/configs/v2.4.2.toml`
  - Normalize `cache_dir` and set `[rpc].cookie_dir = "cache_dir"`
- Documentation updates
  - Add `docs/config-rs-migration-plan.md` describing the new system and precedence
  - Update `generate.rs` comments to show `ZEBRA_SECTION__KEY` env usage

BREAKING CHANGE:
- All legacy `ZEBRA_*` test and path environment variables are removed; use the new names listed above.
- `RUST_LOG`/`ZEBRA_RUST_LOG` are no longer honored; use `ZEBRA_TRACING__FILTER`.
- Docker entrypoint no longer generates or mutates `zebrad.toml`; configuration now relies on defaults, optional TOML, and `ZEBRA_*` env vars.

gustavovalverde committed:
fix: remove extra variable validation

gustavovalverde committed:
Merge branch 'main' into imp-caching

gustavovalverde committed:
Merge branch 'imp-caching' into feat-config-rs

gustavovalverde committed:
chore(deps,deny): bump workspace toml to 0.9; add skip for toml_datetime 0.6.11

- Align with config 0.15.x; dedupe toml-related duplicates
- Add [bans].skip toml_datetime 0.6.11 (via jsonrpsee-proc-macros -> toml_edit 0.22.x); defer jsonrpsee upgrade
- Remove stale/unmatched deny skips; keep ron 0.7.1 and base64 0.21.7 skips

gustavovalverde committed:
fix(test): adapt Docker tests to new config-rs and entrypoint

gustavovalverde committed:
fix: use correct test approach with Docker

gustavovalverde committed:
fix: remaining bans

gustavovalverde committed:
fix(test): correctly log the mining address

gustavovalverde committed:
refactor: remove STATE_CACHE_DIR and LWD_CACHE_DIR from workflows

- ci(sub-ci-integration-tests-gcp): drop STATE_CACHE_DIR and LWD_CACHE_DIR from test_variables; keep TEST_LIGHTWALLETD as the only gate
- tests(acceptance): update comments to no longer require LWD_CACHE_DIR; reference persistent Zebra state via config
- tests(test_type): add default_lwd_cache_dir() and resolve LWD cache path as: env LWD_CACHE_DIR > platform default (if exists) > none
- tests(lightwalletd): clarify LWD_CACHE_DIR docs; note default fallback and speed-up use cases
- docker(entrypoint): remove LWD_CACHE_DIR defaulting; avoid duplicating config
- docs(lightwalletd): document default LWD cache path behavior and optional env override

BREAKING CHANGE:
- STATE_CACHE_DIR is removed from CI and test flows. Use `ZEBRA_STATE__CACHE_DIR` (config-rs) to set `state.cache_dir` when needed.
- LWD_CACHE_DIR is no longer used to gate tests. Tests use a platform default LWD cache directory if present, or an optional `LWD_CACHE_DIR` override for local runs.

gustavovalverde committed:
fix: avoid test filesystem collisions with tests

gustavovalverde committed:
fix: use better mining configuration

gustavovalverde committed:
chore: reduce diff and update docs

gustavovalverde committed:
docs: update CHANGELOG with breaking changes

gustavovalverde committed:
chore: use a single `config.rs` for testing configurations

gustavovalverde committed:
fix: linting

gustavovalverde committed:
fix: simplify the grep command for mining

gustavovalverde committed:
chore: remove flaky test

gustavovalverde committed:
Merge branch 'main' into feat-config-rs

gustavovalverde committed:
Merge branch 'main' into feat-config-rs

gustavovalverde committed:
feat(config): deny sensitive env var overrides via suffix-based filter

- Add is_sensitive_leaf_key() and filter ZEBRA_* vars by leaf key (case-insensitive).
- Block env overrides whose leaf ends with: password, secret, token, cookie, private_key.
- Preserve non-sensitive fields (e.g., paths like cookie_dir, tls_*_path; public values like miner_address).
- Pass sanitized map to config::Environment via .source(Some(...)) with existing prefix/separators.

gustavovalverde committed:
Merge branch 'main' into feat-config-rs

gustavovalverde committed:
fix:

gustavovalverde committed:
fix: SKIP_NETWORK_TESTS for all OSes

gustavovalverde committed:
fix: allow to change the Docker driver for building

gustavovalverde committed:
fix: do not set provenance with default driver

gustavovalverde committed:
fix: provenance default to none

gustavovalverde committed:
fix: disable provenance

gustavovalverde committed:
fix: remove sbom as part of attestation

gustavovalverde committed:
feat(zebrad-config): error on sensitive env vars instead of ignoring

- Fail config loading when sensitive env keys are present (password, secret, token, cookie, private_key)
- Update tests to assert error behavior
- Document security behavior in ZebradConfig::load

arya2 commented on zebra-network/proptest-regressions/peer_set/set/tests/prop.txt:
We can delete this file.

arya2 commented on zebrad/src/application.rs:
```diff
@@ -243,17 +243,27 @@ impl Application for ZebradApp {
+                if error_msg.contains("invalid type: map, expected a string")
+                    || error_msg.contains("eviction_memory_time")
+                {
+                    status_err!("Zebra could not parse the provided config file. This might mean you are using a deprecated format of the file.");
```
Why remove the rest of the log message here? 

Could you add a comment explaining why it logs a different message depending on whether the error message contains the text above?

Update: This has been fixed in https://github.com/ZcashFoundation/zebra/pull/9768/files/7796f788633f537e9294f3051b9be6ca5aa4700c..64aa261dcc2d45c26eff90fc5d47874cfe3e18e6#diff-8f1ee6cceb154001084d1b17ee893fea940caa97b4c2db44f129178845b6e2f8

arya2 commented on zebrad/Cargo.toml:
```diff
@@ -201,6 +201,9 @@ metrics = { workspace = true }
 atty = { workspace = true }
 
+# Configuration management
+config = { version = "0.15.13", features = ["toml", "json"] }
```
Nitpick/Optional: Could we add this as a workspace dependency for consistency?

Update: Done in https://github.com/ZcashFoundation/zebra/pull/9768/commits/d967099ce34d7526bb4afa09a5d39ed6a7fd4bdd

arya2 commented on zebrad/src/commands/copy_state.rs:
```diff
@@ -86,11 +85,11 @@ impl CopyStateCmd {
         let source_config = base_config.state.clone();
 
-        // The default load_config impl doesn't actually modify the app config.
+        // Load the target config using the new config-rs method
```
The `copy-state` command is a little weird since it takes two configs, a source config and a target config.

It probably doesn't make sense to override the config fields in both of them based on the same environment variables. We could accept a prefix in `ZebradConfig::load()` to be expected in the environment variables overriding the target config? Like `TARGET_ZEBRA_STATE__CACHE_DIR` or `ZEBRA_TARGET_STATE__CACHE_DIR`?

Update: Fixed in https://github.com/ZcashFoundation/zebra/pull/9768/commits/a14a50441df9b482818a8c705acb38f2089cc3d0#diff-21d953b03bc82b47d4ced1907ca6e8904561379e2e95f822eb35d9b8a7b6eb88.

<details>
<summary> outdated </summary>
Nitpick/Optional/Subjective:

```suggestion
        // Load the target config if a target config path was provided, or use an ephemeral config otherwise.
```

</details>

arya2 commented on zebrad/src/commands/generate.rs:
```diff
@@ -39,11 +39,22 @@ impl Runnable for GenerateCmd {
+#
+# 2. Configuration file (TOML format)
+#    - Specified via -c flag: `zebrad -c myconfig.toml start`
+#    - Or default location in preference directory (see below)
```
Nitpick/Subjective/Optional:

```suggestion
#    - At the path specified via -c flag, e.g. `zebrad -c myconfig.toml start`, or
#    - At the default path in the user's preference directory (platform-dependent, see below)
```

arya2 commented on zebrad/src/config.rs:
```diff
@@ -54,3 +71,80 @@ pub struct ZebradConfig {
+    /// Configuration is loaded from three sources, in order of precedence:
+    /// 1. Hard-coded defaults (lowest precedence)
+    /// 2. TOML configuration file (if provided)
+    /// 3. Environment variables with `ZEBRA_` prefix (highest precedence)
```
It would be nice to have these in the same order as they are in the `generate` cmd output.

```suggestion
    /// 1. Environment variables with `ZEBRA_` prefix (highest precedence)
    /// 2. TOML configuration file (if provided)
    /// 3. Hard-coded defaults (lowest precedence)
```

arya2 commented on zebrad/src/config.rs:
```diff
@@ -4,8 +4,25 @@
+        || lower.ends_with("cookie")
+        // Only raw private keys; paths like *_private_key_path are not affected.
+        || lower.ends_with("private_key")
+}
```
It would be nice to have these in a constant so they can be referenced below.

```suggestion
/// Centralized, case-insensitive suffix-based deny-list to ban setting config fields with
/// environment variables if those config field names end with any of these suffixes.
const DENY_CONFIG_KEY_SUFFIX_LIST: [&str; 5] = [
    "password",
    "secret",
    "token",
    // Block raw cookies only if a field is literally named "cookie".
    // (Paths like cookie_dir are not affected.)
    "cookie",
    // Only raw private keys; paths like *_private_key_path are not affected.
    "private_key",
];

/// Returns true if a leaf key name should be considered sensitive and blocked
/// from environment variable overrides.
fn is_sensitive_leaf_key(leaf_key: &str) -> bool {
    let key = leaf_key.to_ascii_lowercase();
    DENY_CONFIG_KEY_SUFFIX_LIST.iter().any(|deny_suffix| key.ends_with(deny_suffix))
}
```

arya2 commented on zebrad/src/config.rs:
```diff
@@ -54,3 +71,80 @@ pub struct ZebradConfig {
+    /// # Security
+    /// Environment variables whose leaf key names end with sensitive suffixes (case-insensitive)
+    /// will cause configuration loading to fail with an error: `password`, `secret`, `token`, `cookie`, `private_key`.
+    /// This prevents both silent misconfigurations and process table exposure of sensitive values.
```
```suggestion
    /// # Security
    ///
    /// Environment variables whose leaf key names end with sensitive suffixes (case-insensitive)
    /// will cause configuration loading to fail with an error: `password`, `secret`, `token`, `cookie`, `private_key`.
    /// This prevents both silent misconfigurations and process table exposure of sensitive values.
    ///
    /// See [`DENY_CONFIG_KEY_SUFFIX_LIST`] and [`is_sensitive_leaf_key()`] above
```

arya2 commented on zebrad/src/config.rs:
```diff
@@ -54,3 +71,80 @@ pub struct ZebradConfig {
+    pub fn load(config_path: Option<PathBuf>) -> Result<Self, config::ConfigError> {
+        let mut builder = config::Config::builder();
+
+        // 1. Start with defaults - but don't use try_from with the struct directly
```
It would be nice to specify the struct as `config::Config` in this comment (I initially thought it meant `ZebradConfig`, but that doesn't have a `TryFrom` impl).

It may be clearer to say here that it starts with an empty map and will use the defaults from the `ZebradConfig::default()` impl wherever keys are missing.

Update: Fixed in https://github.com/ZcashFoundation/zebra/pull/9768/commits/baca56046524272d93e4e8c2a19bd0357e72ca97

arya2 commented on zebrad/src/config.rs:
```diff
@@ -4,8 +4,25 @@
 //! application's configuration file and/or command-line options
 //! for specifying it.
 
+use std::path::PathBuf;
```
```suggestion
use std::{collections::HashMap, path::PathBuf};
```

arya2 commented on zebrad/src/config.rs:
```diff
@@ -54,3 +71,80 @@ pub struct ZebradConfig {
+        // We filter the raw environment first, then let config-rs parse types via try_parsing(true).
+        // Sensitive environment variables cause configuration loading to fail for security.
+        let mut filtered_env: std::collections::HashMap<String, String> =
+            std::collections::HashMap::new();
```
```suggestion
        let mut filtered_env: HashMap<String, String> = HashMap::new();
```

arya2 commented on CHANGELOG.md:
```diff
@@ -7,6 +7,30 @@ and this project adheres to [Semantic Versioning](https://semver.org).
+
+### Changed
+
+- `zebrad` now loads configuration from defaults, an optional TOML file, and environment variables,
+  with precedence: Env > TOML > Defaults ([#9768](https://github.com/ZcashFoundation/zebra/pull/9768)).
```
Looks like it allows JSON config files now too.

gustavovalverde replied:
I did not add that just because that's just a "side-effect" of using the `config` crate. But we can add it if we'd like to inform the users about this possibility.

arya2 replied:
I'm happy either way. We can also remove the `json` feature if we don't want to allow it.

arya2 commented on zebrad/src/config.rs:
```diff
@@ -54,3 +71,80 @@ pub struct ZebradConfig {
+            }
+
+            filtered_env.insert(key, value);
+        }
```
Optional/Subjective/Nitpick:

```suggestion
        for (key, value) in std::env::vars().filter(|(key, _)| !key.starts_with("ZEBRA_")) {
            let leaf = key.split("_").last().expect("key must start with 'ZEBRA_'");
            if is_sensitive_leaf_key(leaf) {
                let err_msg = format!(
                    "Environment variable '{key}' contains sensitive key '{leaf}' which \
                     cannot be overridden via environment variables. \
                     Use the configuration file instead to prevent process table exposure.",
                );
                return Err(config::ConfigError::Message(err_msg));
            }

            filtered_env.insert(key, value);
        }
```

gustavovalverde replied:
It seems like this would split on a single underscore (`_`). For a variable like `ZEBRA_RPC__LISTEN_ADDR`, `key.split("_")` would produce `["ZEBRA", "RPC", "", "LISTEN", "ADDR"]`, making the leaf key `"ADDR"` instead of `"LISTEN_ADDR"`

arya2 commented on .config/nextest.toml:
```diff
@@ -7,8 +7,9 @@ status-level = "pass"
 
 # --- All Tests profile ---
 # Runs all tests except 'check_no_git_dependencies'
+# TODO: We must find a better way to remove tests that are ignored, but also require a state or are flaky
```
> Runs all tests except 'check_no_git_dependencies'

This comment needs an update.

Update: Updated in https://github.com/ZcashFoundation/zebra/pull/9768/commits/ca9112f231b7f972fc97a4abc7880a9f970135e1

> We must find a better way to remove tests that are ignored

Do you mean it runs tests that should be ignored like `check_no_git_dependencies`? Is that only the `all-tests` profile, and if so, could we use another profile?

arya2 commented on .github/workflows/sub-build-docker-image.yml:
```diff
@@ -141,7 +141,7 @@ jobs:
         with:
           version: "lab:latest"
-          driver: cloud
+          driver: ${{ vars.DOCKER_BUILDER || 'docker' }}
```
Are we using another driver here?

arya2 commented on .github/workflows/sub-ci-integration-tests-gcp.yml:
```diff
@@ -82,7 +79,7 @@ jobs:
       app_name: zebrad
       test_id: sync-to-mandatory-checkpoint
       test_description: Test sync up to mandatory checkpoint
-      test_variables: "NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=sync-to-mandatory-checkpoint,ZEBRA_TEST_SYNC_TO_CHECKPOINT=1,ZEBRA_CACHE_DIR=/home/zebra/.cache/zebra"
```
Where is the cache directory path being passed in now? It looks like it'll always use the default cache dir path in `test_type.rs` such that stateful tests will no longer be skipped when it's undefined, is that intended?

gustavovalverde replied:
> stateful tests will no longer be skipped when it's undefined

This is correct, this is why I had to add [all these exceptions](https://github.com/ZcashFoundation/zebra/pull/9768/files#diff-3e429424057e0b2a83fafc6e8942f4c05b63f4d977548e51535f8d86be8f9613R12) in the `all-tests` profile in the `nextest.toml`

Maybe this is what I need to consider to improve the comment you refer to in https://github.com/ZcashFoundation/zebra/pull/9768/files#r2292552815


By the way, I wouldn't say it's  completely "intended", but this was the simplest workaround I could think of until we can find a better way to handle stateful tests without adding a feature gate specifically for those tests (as this requires rebuilding Zebra and had a big impact in our testing time)

arya2 replied:
Okay, that seems fine and works. Let's document the decision/change somewhere (maybe the PR description) and we can change it back to checking a constant that either enables or skips the stateful tests if we find a motivation for it.

gustavovalverde replied:
This was initially changed here: https://github.com/ZcashFoundation/zebra/pull/9435 

I added a comment here in the nextest.conf file for reference.

arya2 commented on .github/workflows/sub-ci-integration-tests-gcp.yml:
```diff
@@ -106,7 +103,7 @@ jobs:
       test_id: sync-past-mandatory-checkpoint
       test_description: Test full validation sync from a cached state
-      test_variables: "NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=sync-past-mandatory-checkpoint-${{ (inputs.network || vars.ZCASH_NETWORK) == 'Mainnet' && 'mainnet' || 'testnet' }},ZEBRA_TEST_SYNC_PAST_CHECKPOINT=1,ZEBRA_CACHE_DIR=/home/zebra/.cache/zebra"
+      test_variables: "NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=sync-past-mandatory-checkpoint-${{ (inputs.network || vars.ZCASH_NETWORK) == 'Mainnet' && 'mainnet' || 'testnet' }},TEST_SYNC_PAST_CHECKPOINT=1"
```
Where is `NETWORK` used now? I see it being passed in as `CONTAINER_ENV`, is it meant to define `ZEBRA_NETWORK__NETWORK` somewhere?

gustavovalverde replied:
Oh, this was an oversight on my side. Thank you for catching this. Addressed in [`1a1c5a0` (#9768)](https://github.com/ZcashFoundation/zebra/pull/9768/commits/1a1c5a0ef9045b542caaf14eb420bc2c98ca0120)

arya2 commented on .github/workflows/sub-ci-integration-tests-gcp.yml:
```diff
@@ -295,7 +292,7 @@ jobs:
       app_name: lightwalletd
       test_id: lwd-sync-full
       test_description: Test lightwalletd full sync
-      test_variables: "NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=lwd-sync-full,ZEBRA_TEST_LIGHTWALLETD=1,ZEBRA_CACHE_DIR=/home/zebra/.cache/zebra,LWD_CACHE_DIR=/home/zebra/.cache/lwd"
```
What's replaced `LWD_CACHE_DIR`? Do the tests that were looking for it now always run and use the cached state in the default path if it exists?

gustavovalverde replied:
Yes, that's correct. But the "always" run depends on which nextest profile you're using. This is also related to this answer: https://github.com/ZcashFoundation/zebra/pull/9768/files#r2293155986

arya2 reviewed (commented):
Flushing partial review, looking good so far!

gustavovalverde committed:
Merge branch 'main' into feat-config-rs

gustavovalverde committed:
chore: Apply suggestions from code review

Co-authored-by: Arya <aryasolhi@gmail.com>

gustavovalverde committed:
fix: use `ZEBRA_NETWORK__NETWORK` instead of `NETWORK` in Zebra CI

gustavovalverde committed:
refactor(config): simplify deprecated format checks and error handling

gustavovalverde committed:
chore: better comment

gustavovalverde committed:
chore: enhance comments for `all-tests` profile

gustavovalverde committed:
chore(deps): centralize `config` crate and bump to 0.15.14

- Add `config = { version = "0.15.14", features = ["toml", "json"] }` to `[workspace.dependencies]`
- Switch `zebrad` to `config = { workspace = true }`

gustavovalverde committed:
fix: wrong replace

gustavovalverde committed:
chore: delete extra committed file

gustavovalverde committed:
feat(config): add env-prefix loader and use ZEBRA_TARGET for copy-state

Add `ZebradConfig::load_with_env(config_path, env_prefix)` to allow loading a
config from environment variables using a caller-specified prefix. `load(...)`
delegates to this with the default `ZEBRA` prefix so existing behavior is
unchanged.

Update the `copy-state` command to load its target config with the
`ZEBRA_TARGET_...` prefix (e.g. `ZEBRA_TARGET_STATE__CACHE_DIR`) so source and
target environment overrides do not conflict.

gustavovalverde committed:
fix: PoisonError issues

gustavovalverde committed:
feat(config): add env prefix overrides for `copy-state` command

- Adds `ZebradConfig::load_with_env` to load configs using a specified
  environment variable prefix.
- Updates the `copy-state` command to use the `ZEBRA_TARGET_` prefix for
  its target configuration, preventing conflicts with the source config's
  `ZEBRA_` environment variables.
- Improves test reliability by handling mutex poisoning in `EnvGuard`,
  which prevents a single test failure from cascading.
- Fixes a bug where environment variable overrides were not being correctly
  applied in tests due to incorrect prefix handling.
- Updates documentation and test profiles to reflect these changes.

gustavovalverde committed:
fix: temporarily move some tests out of test-all

arya2 commented:
[This `regtest_block_templates_are_valid_block_submissions()` test failure](https://github.com/ZcashFoundation/zebra/actions/runs/17211645695/job/48827424199?pr=9768#step:4:1161) is due to it using Mainnet instead of Regtest in the spawned test child. The environment variable is overriding the config values written to the test child's temp directory.

[This `downgrade_state_format()` test failure](https://github.com/ZcashFoundation/zebra/actions/runs/17173260709/job/48729546633#step:4:1629) looks like it's passing for Mainnet but failing on Testnet, likely the same issue as it's also writing the config to the test child's temp directory, and as the test is passing for Mainnet.

We probably don't want to define `ZEBRA_NETWORK__NETWORK` in the environment of the `test-all` job.

I think `NETWORK` was previously used in `entrypoint.sh` as part of the base config, so writing a config would override the value, but now it's in the environment and it overrides any value that's written to a config.

arya2 commented on .config/nextest.toml:
```diff
@@ -6,9 +6,22 @@ fail-fast = true
+# TODO: We need a better test architecture to run all non-stateful
 [profile.all-tests]
-default-filter = 'not test(check_no_git_dependencies)'
+default-filter = "not test(check_no_git_dependencies) and not test(=fully_synced_rpc_z_getsubtreesbyindex_snapshot_test) and not test(=lwd_rpc_test) and not test(=lwd_rpc_send_tx) and not test(=lwd_grpc_wallet) and not test(=lwd_integration) and not test(=lwd_sync_full) and not test(=lwd_sync_update) and not test(=lightwalletd_test_suite) and not test(=rpc_get_block_template) and not test(=rpc_su ...
```
I think we should be able to get some of these working

```suggestion
default-filter = "not test(check_no_git_dependencies) and not test(=fully_synced_rpc_z_getsubtreesbyindex_snapshot_test) and not test(=lwd_rpc_test) and not test(=lwd_rpc_send_tx) and not test(=lwd_grpc_wallet) and not test(=lwd_integration) and not test(=lwd_sync_full) and not test(=lwd_sync_update) and not test(=lightwalletd_test_suite) and not test(=rpc_get_block_template) and not test(=rpc_submit_block) and not test(=get_peer_info) and not test(~generate_checkpoints_) and not test(=sync_update_mainnet) and not test(=activate_mempool_mainnet)"
```

gustavovalverde replied:
I merged your PR, so this should be good now

arya2 commented on .config/nextest.toml:
```diff
@@ -103,3 +116,16 @@ default-filter = 'package(zebrad) and test(=lightwalletd_test_suite)'
+
+[profile.nu6-funding-streams-and-coinbase-balance]
+slow-timeout = { period = "30m", terminate-after = 2 }
+default-filter = 'package(zebrad) and test(=nu6_funding_streams_and_coinbase_balance)'
```
```suggestion
```

gustavovalverde replied:
I merged your PR, so this should be good now

arya2 commented on .github/workflows/cd-deploy-nodes-gcp.yml:
```diff
@@ -328,7 +328,7 @@ jobs:
           --container-tty \
           --container-image ${{ vars.GAR_BASE }}/zebrad@${{ needs.build.outputs.image_digest }} \
-          --container-env "NETWORK=${{ matrix.network }},LOG_FILE=${{ vars.CD_LOG_FILE }},SENTRY_DSN=${{ vars.SENTRY_DSN }}" \
+          --container-env "ZEBRA_NETWORK__NETWORK=${{ matrix.network }},LOG_FILE=${{ vars.CD_LOG_FILE }},SENTRY_DSN=${{ vars.SENTRY_DSN }}" \
```
Where is the `ZEBRA_NETWORK__NETWORK` defined here used? It looks like it's only used when deploying the long-running node instances off main or for manual workflows, so I think this one is fine, but just double-checking.

gustavovalverde replied:
This is passed to the Zebra container we deploy through this job. So this is just to specifically set the network Zebra will use (and we commonly deploy the nodes using both Mainnet and Testnet)

arya2 commented on .github/workflows/sub-ci-integration-tests-gcp.yml:
```diff
@@ -82,7 +79,7 @@ jobs:
       test_id: sync-to-mandatory-checkpoint
       test_description: Test sync up to mandatory checkpoint
-      test_variables: "NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=sync-to-mandatory-checkpoint,ZEBRA_TEST_SYNC_TO_CHECKPOINT=1,ZEBRA_CACHE_DIR=/home/zebra/.cache/zebra"
+      test_variables: "ZEBRA_NETWORK__NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=sync-to-mandatory-checkpoint,TEST_SYNC_TO_CHECKPOINT=1"
```
This is running the `sync_to_mandatory_checkpoint_{mainnet, testnet}()` tests, but both will be testing whichever network is defined here.

Is the `ZEBRA_NETWORK__NETWORK` env var defined here expected or being used by this job anywhere?

We could split it up into two jobs if it is expected somewhere.

```suggestion
      test_variables: "NEXTEST_PROFILE=sync-to-mandatory-checkpoint,TEST_SYNC_TO_CHECKPOINT=1"
```

gustavovalverde replied:
I preferred to segregate the tests in the nextest profile [`bcfce06` (#9768)](https://github.com/ZcashFoundation/zebra/pull/9768/commits/bcfce06222f794233abe7aaf9ac51967b09f5b22) to be able to make manual runs of this workflow for Testnet. It was also a bad idea to run this together.

arya2 commented on .github/workflows/sub-ci-integration-tests-gcp.yml:
```diff
@@ -106,7 +103,7 @@ jobs:
       test_id: sync-past-mandatory-checkpoint
       test_description: Test full validation sync from a cached state
-      test_variables: "NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=sync-past-mandatory-checkpoint-${{ (inputs.network || vars.ZCASH_NETWORK) == 'Mainnet' && 'mainnet' || 'testnet' }},ZEBRA_TEST_SYNC_PAST_CHECKPOINT=1,ZEBRA_CACHE_DIR=/home/zebra/.cache/zebra"
+      test_variables: "ZEBRA_NETWORK__NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=sync-past-mandatory-checkpoint-${{ (inputs.network || vars.ZCASH_NETWORK) == 'Mainnet' && 'mainnet' || 'testnet' }},TEST_SYNC_PAST_CHECKPOINT=1"
```
This will only run either the Mainnet or Testnet sync-past-mandatory-checkpoint test depending on `vars.ZCASH_NETWORK`. What was the behaviour before? Are we running both anywhere? Do we want to run both here? If we want to run both here, could we get rid of the `ZEBRA_NETWORK__NETWORK` definition and unify the `sync-past-mandatory-checkpoint-{mainnet, testnet}` profiles?

gustavovalverde replied:
This is expected. We run Mainnet by default, and Testnet when we specifically set to run it. I changed `sync_to_mandatory_checkpoint_{mainnet, testnet}()` to have the same behavior (and thus be consistent).

arya2 commented on .github/workflows/sub-ci-integration-tests-gcp.yml:
```diff
@@ -165,7 +162,7 @@ jobs:
       test_id: sync-update-mainnet
       test_description: Test syncing to tip with a Zebra tip state
-      test_variables: "NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=sync-update-mainnet,ZEBRA_CACHE_DIR=/home/zebra/.cache/zebra"
+      test_variables: "ZEBRA_NETWORK__NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=sync-update-mainnet"
```
We could probably remove the `ZEBRA_NETWORK__NETWORK` definition here altogether, but, if we're keeping it, the `needs.sync-full-mainnet.result == 'success'` condition above implies this will always be Mainnet.

```suggestion
      test_variables: "ZEBRA_NETWORK__NETWORK=Mainnet,NEXTEST_PROFILE=sync-update-mainnet"
```

Update: Not a blocker.

arya2 commented on .github/workflows/sub-ci-integration-tests-gcp.yml:
```diff
@@ -422,7 +419,7 @@ jobs:
       test_id: rpc-get-block-template
       test_description: Test getblocktemplate RPC method via Zebra's rpc server
-      test_variables: "NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=rpc-get-block-template,ZEBRA_CACHE_DIR=/home/zebra/.cache/zebra"
+      test_variables: "ZEBRA_NETWORK__NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=rpc-get-block-template"
```
The `needs.sync-full-mainnet.result == 'success'` condition implies this will only run on Mainnet, the test itself hard codes Mainnet but should work fine on Testnet too if it has an up-to-date Testnet cache.

```suggestion
      test_variables: "NEXTEST_PROFILE=rpc-get-block-template"
```

Update: Not a blocker.

arya2 commented on .github/workflows/sub-ci-integration-tests-gcp.yml:
```diff
@@ -445,7 +442,7 @@ jobs:
       test_id: rpc-submit-block
       test_description: Test submitting blocks via Zebra's rpc server
-      test_variables: "NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=rpc-submit-block,ZEBRA_CACHE_DIR=/home/zebra/.cache/zebra"
+      test_variables: "ZEBRA_NETWORK__NETWORK=${{ inputs.network || vars.ZCASH_NETWORK }},NEXTEST_PROFILE=rpc-submit-block"
```
This test was always running on Mainnet before, now it could run on either depending on which network is defined here. It should run fine on Testnet if this env var configures it to do so.

gustavovalverde replied:
Yes, this is intended.

arya2 commented on .github/workflows/sub-ci-unit-tests-docker.yml:
```diff
@@ -52,7 +52,7 @@ jobs:
           docker run --tty \
           -e NEXTEST_PROFILE=all-tests \
-          -e NETWORK="${{ inputs.network || vars.ZCASH_NETWORK }}" \
+          -e ZEBRA_NETWORK__NETWORK="${{ inputs.network || vars.ZCASH_NETWORK }}" \
```
This is causing some failures in tests that cover both Mainnet and Testnet, or which use Regtest or a configured Testnet. If it's necessary to define the env var here, we could define another prefix for base config values that can be overridden by the config file, define a new env var or CLI option to express that the config file shouldn't be overridden, or find some other workaround, but removing it is an easy fix otherwise.

```suggestion
```

arya2 commented on .github/workflows/sub-ci-unit-tests-docker.yml:
```diff
@@ -74,14 +74,12 @@ jobs:
-          -e ZEBRA_TEST_LARGE_CHECKPOINTS=1 \
-          -e NETWORK="${{ inputs.network || vars.ZCASH_NETWORK }}" \
+          -e TEST_LARGE_CHECKPOINTS=1 \
+          -e ZEBRA_NETWORK__NETWORK="${{ inputs.network || vars.ZCASH_NETWORK }}" \
```
This test was always running on Mainnet before, but should work fine on Testnet as well.

gustavovalverde replied:
Yes, this is intended.

arya2 commented on .github/workflows/sub-ci-unit-tests-docker.yml:
```diff
@@ -103,14 +101,12 @@ jobs:
-          -e ZEBRA_TEST_LIGHTWALLETD=1 \
-          -e NETWORK="${{ inputs.network || vars.ZCASH_NETWORK }}" \
+          -e TEST_LIGHTWALLETD=1 \
+          -e ZEBRA_NETWORK__NETWORK="${{ inputs.network || vars.ZCASH_NETWORK }}" \
```
This test was always running on Mainnet before, but may work fine on Testnet as well.

gustavovalverde replied:
Yes, this is intended.

arya2 commented on .github/workflows/sub-ci-unit-tests-docker.yml:
```diff
@@ -163,7 +159,7 @@ jobs:
           docker run --tty \
           -e NEXTEST_PROFILE=check-no-git-dependencies \
-          -e NETWORK="${{ inputs.network || vars.ZCASH_NETWORK }}" \
+          -e ZEBRA_NETWORK__NETWORK="${{ inputs.network || vars.ZCASH_NETWORK }}" \
```
This test doesn't use the network anywhere, it's just reading the `Cargo.lock` file and checking that there are no git sources in there.

gustavovalverde replied:
I'll consider it in the CI refactor.

arya2 commented on zebrad/src/config.rs:
```diff
@@ -4,8 +4,32 @@
 
+/// Centralized, case-insensitive suffix-based deny-list to ban setting config fields with
+/// environment variables if those config field names end with any of these suffixes.
+const DENY_CONFIG_KEY_SUFFIX_LIST: [&str; 5] = [
```
Should the `miner_address` be included here? Or the RPC listen addr? There's a DoS vulnerability if the RPC listen address is ever publicly exposed (even with a cookie, someone could repeatedly call the server and starve Zebra).

Update: Added to an issue.

arya2 commented on .github/workflows/sub-test-zebra-config.yml:
```diff
@@ -29,54 +35,54 @@ jobs:
-            grep_patterns: '-e "miner_address = \\\"u1cymdny2u2vllkx7t5jnelp0kde0dgnwu0jzmggzguxvxj6fe7gpuqehywejndlrjwgk9snr6g69azs8jfet78s9zy60uepx6tltk7ee57jlax49dezkhkgvjy2puuue6dvaevt53nah7t2cc2k4p0h0jxmlu9sx58m2xdm5f9sy2n89jdf8llflvtml2ll43e334avu2fwytuna404a\\\""'
+            env_vars: -e FEATURES=prometheus -e ZEBRA_METRICS__ENDPOINT_ADDR=0.0.0.0:9999
+            grep_patterns: -e "0.0.0.0:9999"
+          # Todo: This test is flaky, and we need to fix it. But this is included in the Rust tests as a fallback.
```
It seems strange that this test would be flaky, it looks like the network would always be Mainnet, I would expect it to consistently pass. In any case, the config tests should fail if Zebra can't parse a valid miner address, so it doesn't seem necessary to check that here too.

gustavovalverde replied:
I'll leave it here to have it as a reference in my refactor. But yes, the test in the config should be enough in any case.

arya2 commented on Cargo.toml:
```diff
@@ -132,10 +132,10 @@ tokio = "1.44"
 tonic = "0.14"
 tonic-prost = "0.14"
-tonic-prost-build = "0.14"
+tonic-prost-build = "0.14" 
```
Nitpick/Optional:

```suggestion
tonic-prost-build = "0.14"
```

arya2 commented on Cargo.toml:
```diff
@@ -153,6 +153,7 @@ wagyu-zcash-parameters = "0.2"
 x25519-dalek = "2.0.1"
 zcash_note_encryption = "0.4.1"
 zcash_script = "0.3.2"
+config = { version = "0.15.14", features = ["toml", "json"] }
```
Was the `json` feature selected intentionally? As far as I can tell it just allows JSON config files, I see no harm in allowing that if we want to keep this, I'm just curious why it was added.

gustavovalverde replied:
We don't have an use-case, so I'll remove it [chore: remove json feature from config](https://github.com/ZcashFoundation/zebra/pull/9768/commits/1cb799bf3234220e369887e6c5a99d0e111b8379)

arya2 commented on book/src/user/docker.md:
```diff
@@ -80,28 +80,26 @@ All available Cargo features are listed at
 ## Configuring Zebra
 
-To configure Zebra using Docker, you have a few options, processed in this order:
+Zebra uses config-rs to layer configuration from defaults, an optional TOML file, and `ZEBRA_`-prefixed environment variables. When running with Docker, configure Zebra using any of the following (later items override earlier ones):
```
```suggestion
Zebra uses [config-rs](https://crates.io/crates/config) to layer configuration from defaults, an optional TOML file, and `ZEBRA_`-prefixed environment variables. When running with Docker, configure Zebra using any of the following (later items override earlier ones):
```

arya2 commented on book/src/user/docker.md:
```diff
@@ -80,28 +80,26 @@ All available Cargo features are listed at
 ### RPC
 
-Zebra's RPC server is disabled by default. To enable it, you need to define the RPC settings in Zebra's configuration. You can achieve this using one of the configuration methods described above:
+Zebra's RPC server is disabled by default. Enable and configure it via TOML or environment variables:
```
Nitpick/Optional:

```suggestion
Zebra's RPC server is disabled by default. Enable and configure it via the TOML configuration file, or configuration environment variables:
```

arya2 commented on book/src/user/docker.md:
```diff
@@ -137,11 +135,9 @@ docker compose -f docker/docker-compose.lwd.yml up
-configured the RPC server by setting environment variables directly in the
-`docker/docker-compose.lwd.yml` file. This takes advantage of the entrypoint
-script's auto-generation feature (method 3 described above).
+authentication mechanism disabled since Lightwalletd doesn't support it. In this
```
Nitpick: It would be nice to be more specific here.

```suggestion
authentication mechanism disabled when running `docker compose -f docker/docker-compose.lwd.yml up`, since Lightwalletd doesn't support cookie authentication. In this
```

arya2 commented on docker/entrypoint.sh:
```diff
@@ -2,24 +2,12 @@
-# - `$ZEBRA_CONF_PATH` can point to an existing Zebra config file, or if not set,
-#   the script will look for a default config at ${HOME}/.config/zebrad.toml,
-#   or generate one using environment variables.
+# This script handles privilege dropping and launches zebrad or tests.
```
Do we want to mention `CONFIG_FILE_PATH` here where it mentioned `ZEBRA_CONF_PATH` before?

arya2 commented on docker/entrypoint.sh:
```diff
@@ -31,86 +19,13 @@ exec_as_user() {
 }
 
 # Creates a directory if it doesn't exist and sets ownership to specified UID:GID.
-# Also ensures the parent directories have the correct ownership.
-#
-# ## Parameters
-#
-# - $1: Directory path to create and own
```
Why remove these docs?

arya2 commented on docker/entrypoint.sh:
```diff
@@ -130,64 +45,30 @@ create_owned_directory() {
-  fi
+# If the user provided a config file path we pass it to zebrad.
+CONFIG_ARGS=()
+if [[ -n ${CONFIG_FILE_PATH} && -f ${CONFIG_FILE_PATH} ]]; then
```
Do we want to still print a warning if a file can't be found at the provided path?

arya2 commented on docker/entrypoint.sh:
```diff
@@ -130,64 +45,30 @@ create_owned_directory() {
+# Main Script Logic
 # - If "$1" is "--", "-", or "zebrad", run `zebrad` with the remaining params.
-# - If "$1" is "test":
-#   - and "$2" is "zebrad", run `zebrad` with the remaining params,
```
Was this wrong before? I see that it was checking `"$1" == "zebrad"` before this change as well, why?

arya2 commented on docker/entrypoint.sh:
```diff
@@ -130,64 +45,30 @@ create_owned_directory() {
   if [[ "$1" == "zebrad" ]]; then
     shift
-    exec_as_user zebrad --config "${ZEBRA_CONF_PATH}" "$@"
+    exec_as_user zebrad "${CONFIG_ARGS[@]}" "$@"
```
Will it ever get here? It looks like the case statement above will always catch it before it does?

Update: Not a blocker.

arya2 commented on zebrad/src/application.rs:
```diff
@@ -243,17 +243,19 @@ impl Application for ZebradApp {
+            // Ignore errors loading the config for some commands.
+            Err(_e) if command.cmd().should_ignore_load_config_error() => Default::default(),
+            Err(e) => {
+                status_err!("Zebra could not load its configuration: {}", e);
```
It would be nice to leave the extra guidance here.

```suggestion
                status_err!(
                    "Zebra could not load the provided configuration file and/or environment variables.\
                     This might mean you are using a deprecated format of the file, or are attempting to
                     configure deprecated or unknown fields via environment variables.\
                     You can generate a valid config by running \"zebrad generate\", \
                     and diff it against yours to examine any format inconsistencies."
                );
```

arya2 commented on zebrad/src/config.rs:
```diff
@@ -54,3 +78,93 @@ pub struct ZebradConfig {
+                let parts: Vec<&str> = without_prefix.split("__").collect();
+                if let Some(leaf) = parts.last() {
+                    if is_sensitive_leaf_key(leaf) {
+                        return Err(config::ConfigError::Message(format!(
+                            "Environment variable '{}' contains sensitive key '{}' which cannot be overridden via environment variables. \
+                             Use the configuration file instead to prevent process table exposure.",
+                            key, leaf
+                        )));
```
Why return an error here rather than logging a warning and ignoring the sensitive key?

gustavovalverde replied:
I changed it based on this comment in Zallet https://github.com/zcash/wallet/pull/194#discussion_r2288082278

arya2 commented on zebrad/tests/acceptance.rs:
```diff
@@ -30,25 +30,25 @@
-//!   created in the file system with write permissions.
+//! - `ZEBRA_STATE__CACHE_DIR` env variable: The path to a Zebra cached state directory.
+//!   For some sync tests, this directory needs to be created in the file system
+//!   with write permissions.
```
```suggestion
//! - A zebrad state cache directory is required for some tests, either at the default state cache
//!   directory path, or at the path defined in the `ZEBRA_STATE__CACHE_DIR` env variable.
//!   For some sync tests, this directory needs to be created in the file system
//!   with write permissions.
```

arya2 commented on .config/nextest.toml:
The 3 week timeout for `sync-full-mainnet` seems excessive, do we want to change that to 1 week? Wasn't changed in this PR if we want to leave it for now.

gustavovalverde replied:
I'll change it in the CI refactor.

arya2 commented on zebrad/tests/acceptance.rs:
Adding `#[cfg(not(target_os = "windows"))]` above `fn last_config_is_stored()` should fix the lint.

arya2 commented on zebrad/tests/acceptance.rs:
```diff
@@ -903,7 +903,7 @@ fn invalid_generated_config() -> Result<()> {
 
     // Check that Zebra produced an informative message.
-    output.stderr_contains("Zebra could not parse the provided config file. This might mean you are using a deprecated format of the file.")?;
+    output.stderr_contains("Zebra could not load its configuration: invalid type: map, expected a string for key `mempool.eviction_memory_time`")?;
```
If you commit the suggestion to change this error message in `application.rs`, this will need an update too. Otherwise, this should still be updated to accept any error message that includes the first hard-coded part (so, excluding the part after the colon).

```suggestion
    output.stderr_contains("Zebra could not load the provided configuration file and/or environment variables")?;
```

arya2 commented on zebrad/tests/acceptance.rs:
```diff
@@ -1780,7 +1784,8 @@ fn lwd_integration() -> Result<()> {
 /// Make sure `zebrad` can sync from peers, but don't actually launch `lightwalletd`.
 ///
-/// This test only runs when the `ZEBRA_CACHE_DIR` env var is set.
+/// This test only runs when a persistent cached state is configured
```
This and similar updates below look wrong, because it'll use the default state cache dir path if none is configured.

Nitpick:

```suggestion
/// This test only runs when a persistent cached state directory path is configured
```

arya2 commented on zebrad/tests/acceptance.rs:
```diff
@@ -1791,8 +1796,8 @@ fn sync_update_mainnet() -> Result<()> {
-/// - the `ZEBRA_TEST_LIGHTWALLETD`, `ZEBRA_CACHE_DIR`, and
-///   `LWD_CACHE_DIR` env vars are set, and
+/// - `TEST_LIGHTWALLETD` is set,
+/// - a persistent cached state is configured (e.g., via `ZEBRA_STATE__CACHE_DIR`), and
```
Nitpick:

```suggestion
/// - a persistent cached state directory path is configured (e.g., via `ZEBRA_STATE__CACHE_DIR`), and
```

arya2 commented on zebrad/tests/acceptance.rs:
```diff
@@ -1825,9 +1831,9 @@ fn lwd_sync_full() -> Result<()> {
 /// Runs the tests in this order:
 /// - launch lightwalletd with empty states,
-/// - if `ZEBRA_CACHE_DIR` is set:
+/// - if a cached Zebra state is configured:
```
```suggestion
/// - if a cached Zebra state directory path is configured:
```

arya2 commented on zebrad/tests/acceptance.rs:
```diff
@@ -1825,9 +1831,9 @@ fn lwd_sync_full() -> Result<()> {
+/// - if a cached Zebra state is configured:
 ///   - run a full sync
-/// - if `ZEBRA_CACHE_DIR` and `LWD_CACHE_DIR` are set:
+/// - if a cached Zebra state is configured:
```
```suggestion
/// - if a cached Zebra state directory path is configured:
```

arya2 commented on zebrad/tests/acceptance.rs:
```diff
@@ -1843,29 +1849,29 @@ async fn lightwalletd_test_suite() -> Result<()> {
     })?;
 
-    // Only runs when ZEBRA_CACHE_DIR is set.
+    // Only runs when a cached Zebra state is configured.
```
```suggestion
    // Only runs when a cached Zebra state directory path is configured with an environment variable.
```

arya2 commented on zebrad/tests/common/cached_state.rs:
```diff
@@ -31,8 +31,8 @@ use crate::common::{
-/// The environmental variable that holds the path to a directory containing a cached Zebra state.
-pub const ZEBRA_CACHE_DIR: &str = "ZEBRA_CACHE_DIR";
+// Deprecated: tests should read the cache dir from `ZEBRA_STATE__CACHE_DIR` if they need
+// an external cached state directory. The actual application reads this via config-rs.
```
```suggestion
```

arya2 commented on zebrad/tests/common/config.rs:
```diff
@@ -53,33 +53,41 @@ pub fn default_test_config(net: &Network) -> Result<ZebradConfig> {
-        env::var("ZEBRA_FORCE_USE_COLOR"),
-        Err(env::VarError::NotPresent)
-    );
+    let force_use_color = env::var("FORCE_USE_COLOR").is_ok();
```
I'm just now noticing this, it doesn't appear to be documented anywhere if it could be added to the changelog or somewhere in this PR while you're here (it wasn't documented anywhere before either).

arya2 commented on zebrad/tests/common/config.rs:
```diff
@@ -53,33 +53,41 @@ pub fn default_test_config(net: &Network) -> Result<ZebradConfig> {
+    let effective_kind: NetworkKind = match env::var("ZEBRA_NETWORK__NETWORK").ok().as_deref() {
+        Some("Mainnet") | Some("mainnet") | Some("Main") | Some("main") => NetworkKind::Mainnet,
+        Some("Regtest") | Some("regtest") => NetworkKind::Regtest,
+        Some("Testnet") | Some("testnet") => NetworkKind::Testnet,
```
Lowercase/partial values aren't parsed as a valid `NetworkKind` (we could change this if we want, but, it's just not the case currently, and Zebra will fail to load the configuration from the environment if they're provided).

```suggestion
        Some("Mainnet") => NetworkKind::Mainnet,
        Some("Regtest") => NetworkKind::Regtest,
        Some("Testnet") => NetworkKind::Testnet,
```

arya2 commented on zebrad/tests/common/config.rs:
The changes here may cause some unexpected behaviour, because, until now, tests could assume that the config passed to the test zebrad child would be the one that it will use, but with the changes in this PR, they could be overridden with environment variables.

`trusted_chain_sync_handles_forks_correctly()` passes a config struct directly to `TrustedChainSync::spawn()` and also passes it to `with_config()` when spawning a test zebrad child, and there may be other instances where this is relevant.

We should document this behaviour and update `ZebradConfig` to disallow instantiation without going through the layered construction that will be used when loading the config in `application.rs` (perhaps as a separate issue since this PR is already quite large?).

arya2 commented on zebrad/tests/common/sync.rs:
```diff
@@ -326,14 +326,14 @@ pub fn check_sync_logs_until(
 }
 
-/// Returns the cache directory for Zebra's state.
+/// Returns the cache directory for Zebra's state, as configured via config-rs.
```
```suggestion
/// Returns the cache directory for Zebra's state, as configured with [`ZebradConfig::load`] with config-rs.
```

arya2 commented on zebrad/tests/common/sync.rs:
```diff
@@ -326,14 +326,14 @@ pub fn check_sync_logs_until(
-/// It checks the `ZEBRA_CACHE_DIR` environment variable and returns its value if set.
-/// Otherwise, it defaults to `"/zebrad-cache"`.
+/// Uses the resolved configuration from `ZebradConfig::load(None)`, which
+/// incorporates defaults, optional TOML, and environment overrides.
```
```suggestion
/// Uses the resolved configuration from [`ZebradConfig::load(None)`](ZebradConfig::load), which
/// incorporates defaults, optional TOML, and environment overrides.
```

arya2 commented on zebrad/tests/common/test_type.rs:
```diff
@@ -152,18 +183,29 @@ impl TestType {
-                let test_name = test_name.as_ref();
+        // Load the effective config (defaults + optional TOML + env overrides)
+        let test_name = test_name.as_ref();
+        let cfg = match ZebradConfig::load(None) {
```
We should either update the documentation for acceptance tests that required a cached state to say that they'll always run and look for a cached state in the default cache state directory path, or update this to check for the environment variable outside of `ZebradConfig::load()` so that it won't use the default value if no value is configured.

arya2 commented on zebrad/tests/common/test_type.rs:
```diff
@@ -241,23 +283,38 @@ impl TestType {
+            if default_path.exists() {
+                Some(default_path)
+            } else {
+                tracing::info!(
```
```suggestion
                tracing::warn!(
```

arya2 commented on zebrad/tests/common/test_type.rs:
```diff
@@ -241,23 +283,38 @@ impl TestType {
+            if default_path.exists() {
+                Some(default_path)
+            } else {
+                tracing::info!(
```
```suggestion
                tracing::warn!(
```

arya2 commented on zebrad/tests/config.rs:
```diff
@@ -0,0 +1,342 @@
+        let guard = TEST_MUTEX.lock().unwrap_or_else(|e| e.into_inner());
+
+        let original_vars: Vec<(String, String)> = env::vars()
+            .filter(|(key, _val)| key.starts_with("ZEBRA_"))
```
Could we move this prefix to a constant and use the constant here and in the `Drop` impl below?

gustavovalverde replied:
Solved here [`594874a` (#9768)](https://github.com/ZcashFoundation/zebra/pull/9768/commits/594874a50f00b4cf1f20e8dddeda95007b012ebe)

arya2 commented on zebrad/tests/config.rs:
```diff
@@ -0,0 +1,342 @@
+
+    let nonexistent_path = PathBuf::from("/this/path/does/not/exist.toml");
+    let result = ZebradConfig::load(Some(nonexistent_path));
+    assert!(result.is_err(), "Should fail to load nonexistent file");
```
Optional/Nitpick: We can use `.expect_err()` here and below where we don't want to match on a specific error value.

```suggestion
    let _env = EnvGuard::new();
    ZebradConfig::load(Some(PathBuf::from("/this/path/does/not/exist.toml")))
        .expect_err("should fail to load nonexistent file");
```

gustavovalverde replied:
Solved here, for multiple cases [`594874a` (#9768)](https://github.com/ZcashFoundation/zebra/pull/9768/commits/594874a50f00b4cf1f20e8dddeda95007b012ebe)

arya2 commented on zebrad/tests/config.rs:
```diff
@@ -0,0 +1,342 @@
+
+    let config = ZebradConfig::load(None).expect("load config with nested env vars");
+
+    assert_eq!(config.tracing.filter.as_ref().unwrap(), "debug");
```
Optional/Nitpick: It would be nice to assert that these are some values instead of calling `.unwrap()`.

```suggestion
    assert_eq!(config.tracing.filter.as_ref(), Some("debug"));
```

arya2 commented on zebrad/tests/config.rs:
```diff
@@ -0,0 +1,342 @@
+}
+
+#[test]
+fn zebra_mining_miner_address_from_toml() {
```
Optional/Nitpick: It would be nice to prefix these test names with `config_` 

```suggestion
fn config_zebra_network_network_env() {
    let env = EnvGuard::new();

    env.set_var("ZEBRA_NETWORK__NETWORK", "Testnet");

    let config = ZebradConfig::load(None).expect("load config with ZEBRA_NETWORK__NETWORK");
    assert_eq!(config.network.network.to_string(), "Testnet");
}

#[test]
fn config_zebra_rpc_listen_addr_env() {
    let env = EnvGuard::new();

    env.set_var("ZEBRA_RPC__LISTEN_ADDR", "127.0.0.1:18232");

    let config = ZebradConfig::load(None).expect("load config with ZEBRA_RPC__LISTEN_ADDR");
    assert_eq!(
        config.rpc.listen_addr.unwrap().to_string(),
        "127.0.0.1:18232"
    );
}

#[test]
fn config_zebra_state_cache_dir_env() {
    let env = EnvGuard::new();

    env.set_var("ZEBRA_STATE__CACHE_DIR", "/test/cache");

    let config = ZebradConfig::load(None).expect("load config with ZEBRA_STATE__CACHE_DIR");
    assert_eq!(config.state.cache_dir, PathBuf::from("/test/cache"));
}

#[test]
fn config_zebra_metrics_endpoint_addr_env() {
    let env = EnvGuard::new();

    env.set_var("ZEBRA_METRICS__ENDPOINT_ADDR", "0.0.0.0:9999");

    let config = ZebradConfig::load(None).expect("load config with ZEBRA_METRICS__ENDPOINT_ADDR");
    assert_eq!(
        config.metrics.endpoint_addr.unwrap().to_string(),
        "0.0.0.0:9999"
    );
}

#[test]
fn config_zebra_tracing_log_file_env() {
    let env = EnvGuard::new();

    env.set_var("ZEBRA_TRACING__LOG_FILE", "/test/zebra.log");

    let config = ZebradConfig::load(None).expect("load config with ZEBRA_TRACING__LOG_FILE");
    assert_eq!(
        config.tracing.log_file.as_ref().unwrap(),
        &PathBuf::from("/test/zebra.log")
    );
}

#[test]
fn config_zebra_mining_miner_address_from_toml() {
```

arya2 commented on zebrad/tests/config.rs:
```diff
@@ -0,0 +1,342 @@
+    let toml_string = format!(
+        "[network]\nnetwork = \"Testnet\"\n\n[mining]\nminer_address = \"{}\"",
+        miner_address
+    );
```
Optional/Nitpick: We could use a raw str here to make this more readable.

```suggestion
    let toml_string = format!(
        r#"[network]
        network = "Testnet"
        
        [mining]
        miner_address = "{miner_address}""#,
    );
```

arya2 commented on zebrad/tests/config.rs:
```diff
@@ -0,0 +1,342 @@
+    {
+        let msg = result.unwrap_err().to_string();
+        assert!(msg.contains("sensitive key"), "error message: {}", msg);
+    }
```
This should work regardless of if the feature is selected.

```suggestion
    let msg = result.unwrap_err().to_string();
    assert!(msg.contains("sensitive key"), "error message: {}", msg);
```

arya2 reviewed (commented):
This is looking good, thank you for making all these changes.

arya2 committed:
Removes `ZEBRA_NETWORK__NETWORK` env var from `Run all tests` (#9830)

* Removes `ZEBRA_NETWORK__NETWORK` env var from `Run all tests`

* removes duplicate 'and'

gustavovalverde committed:
chore: apply suggestions from code review

Co-authored-by: Arya <aryasolhi@gmail.com>

gustavovalverde committed:
fix: format and clippy

gustavovalverde committed:
fix(tests): make `sync-to-mandatory-checkpoint` network-specific

Split `sync-to-mandatory-checkpoint` nextest profile into
`sync-to-mandatory-checkpoint-mainnet` and
`sync-to-mandatory-checkpoint-testnet` so each profile runs a single
network-specific test.

Update CI job `sub-ci-integration-tests-gcp.yml`
to select `NEXTEST_PROFILE` dynamically per network so only the matching
profile runs in a workflow invocation. This prevents concurrent tests for
different networks from sharing the same cache dir and creating race conditions.

gustavovalverde committed:
chore: remove `json` feature from `config`

gustavovalverde committed:
chore(test): add ZEBRA_ENV_PREFIX and use expect_err in tests

- Add `ZEBRA_ENV_PREFIX` const and use it in `EnvGuard::new` and the `Drop` impl.
- Replace `assert!(result.is_err())` with `.expect_err(...)` in tests that don't inspect the error value.

gustavovalverde committed:
Merge branch 'main' into feat-config-rs

gustavovalverde committed:
fix: cargo.lock merge conflict

gustavovalverde committed:
chore: suggestions from code review

Co-authored-by: Arya <aryasolhi@gmail.com>

gustavovalverde committed:
chore: format

arya2 commented on zebrad/src/application.rs:
```diff
@@ -248,7 +248,13 @@ impl Application for ZebradApp {
+                     This might mean you are using a deprecated format of the file, or are attempting to
+                     configure deprecated or unknown fields via environment variables.\
+                     You can generate a valid config by running \"zebrad generate\", \
+                     and diff it against yours to examine any format inconsistencies."
```
```suggestion
                     and diff it against yours to examine any format inconsistencies. Error: {e}"
```

arya2 reviewed (approved):
These changes look great, thank you!

Opened https://github.com/ZcashFoundation/zebra/issues/9842 to track non-blocking comments that we should revisit.

arya2 commented:
@Mergifyio refresh

gustavovalverde commented:
I'll admin merge to be able to clean the commit message when the PR merges
